### PR TITLE
DocsScraper sub repo

### PR DESCRIPTION
**Pull request by splendidbug**
This is a new subrepo dedicated to crawling, scraping, parsing and chunking of Julia's documentation (GSOC project).

DocsScraper contains DocsScraper.jl file where the code for parser is implemented. 

Usage: 
`parsed_text = parse_url("https://docs.julialang.org/en/v1/base/multi-threading/")`

Returns: 
A Vector of Dict containing Heading/Text/Code along with a Dict of respective metadata

Requirements:
HTTP, Gumbo, AbstractTrees, URIs

### CODE REVIEW

**Comment by svilupp** [comment]
Great start! It's very clear.

I just had a couple of stylistic comments - let me know what you think.

**Comment by svilupp** [comment]
It would also be good to add some minimal tests in folder `test/` -> the entry point is usually file `test/runtests.jl` and you'll need to add [extras] to your Project.toml - see the parent project for example.

It would be good to cover at least the main pathways of the core functions

**Review by svilupp — CHANGES_REQUESTED** [review]

File: DocsScraper/src/DocsScraper.jl [hunk 1]
```diff
@@ -0,0 +1,152 @@
+"""
+    get_base_url(url::String)
+
+Extracts the base url.
+
+# Arguments
+- `url`: The url string of which, the base url needs to be extracted
+"""
+function get_base_url(url::String)
```
[the remark this answers is not in the document]
Comment (splendidbug) [reply DocsScraper/src/DocsScraper.jl:9]: Insightful!

File: DocsScraper/src/DocsScraper.jl [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+- `node`: The root HTML node 
+- `heading_hierarchy`: Dictionary used to store metadata
+- `parsed_text`: Vector of Dicts to store parsed text and metadata
+
+# TODO:
+- Inline code blocks are the ones present inside <p> or <li>. put `` around the inline code blocks. Make changes inside `elseif tag_name == "p" || tag_name == "li"`
+"""
+function process_nodes(node, heading_hierarchy::Dict, parsed_text::Vector)
+    # if a node is an html element
+    if typeof(node) <: Gumbo.HTMLElement
```
[the remark this answers is not in the document]
Comment (splendidbug) [reply DocsScraper/src/DocsScraper.jl:31]: Good idea. If I do this then, `function process_nodes(node::Gumbo.HTMLText, ...) ... end ` doesn't require other parameters: `heading_hierarchy`, `parsed_text`. so to maintain clarity, I can do any one from the following:
1) Leave the unused `heading_hierarchy` and `parsed_text`
2) remove `heading_hierarchy` and `parsed_text`. If I do this, then I have to add an if condition while function calls
3) use args... with proper comments to represent the data structures passed.

Which method do you think would be the best?

File: DocsScraper/src/DocsScraper.jl [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+            # if the current tag is <code>, then get it's textual value and store in the dictionary with key as "code"
+        elseif tag_name == "code"
+            # Start a new code block
+            code_content = strip(Gumbo.text(node))
+            push!(parsed_text, Dict("code" => code_content, "metadata" => copy(heading_hierarchy)))
+            return ""
+
+            # if the current tag is <p> or <li>, then store the text of the whole tag 
+        elseif tag_name == "p" || tag_name == "li"
```
[the remark this answers is not in the document]
Comment (splendidbug) [reply DocsScraper/src/DocsScraper.jl:57]: Yes, there are other tags like span, ul, ol, table, etc., but if any other tag contains text, then it'll eventually end up inside HTMLText (since we're going deeper into html hierarchy in every iteration), so there's no need to explicitly handle the others

**Comment by svilupp** [comment]
@splendidbug How are you thinking about the scope of this PR before we merge it?

I think we can have it open for a while, but I'd like to see the following prior to merge:
- functionality:
  - download web page (can be in memory, or saved to a local file for future reference/re-parsing, it should reflect the webpage structure. HTML file only)
  - parsing HTML page
  - extracting links from HTML page 
  - filtering pages to only the same domain(?, should be an argument
  - top-level functionality to run the scrape->links->parse & repeat on other links in scope
  
- tests for the main behaviours

It doesn't have to cover all edge cases, but it would be good to cover all these steps, because it will force us to design a good interface/API from the start (and not change it soon).

Eg, in separate PRs we should consider checking anti-scraping measures, eg, robots.txt, and make sure we adhere to it. But that's extra and not the core scope of this PR.

WDYT?

File: DocsScraper/src/DocsScraper.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (svilupp) [reply DocsScraper/src/DocsScraper.jl:31]: If you want to keep the same interface, 1 and 3 are possible. 

1) is slightly more strict so potentially more robust to accidental bugs in the future (and can be easily changed if needed), but 3) is correct as well.

I have done both. I think I tend to default to 3) because of convenience

[the remark this answers is not in the document]
Comment (splendidbug) [reply DocsScraper/src/DocsScraper.jl:31]: Gotch. Thanks!

File: DocsScraper/src/DocsScraper.jl [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+# Arguments
+- `node`: The root HTML node 
+- `heading_hierarchy`: Dictionary used to store metadata
+- `parsed_text`: Vector of Dicts to store parsed text and metadata
+
+# TODO:
+- Inline code blocks are the ones present inside <p> or <li>. put `` around the inline code blocks. Make changes inside `elseif tag_name == "p" || tag_name == "li"`
+"""
+function process_nodes(node, heading_hierarchy::Dict, parsed_text::Vector)
```
[the remark this answers is not in the document]
Comment (splendidbug) [reply DocsScraper/src/DocsScraper.jl:29]: > What's the thinking behind providing `heading_hierarchy` and `parsed_text`? Is it to incrementally add to the same objects when scanning multiple URLs?
> 
> In general, unless there are huge performance benefits, it will be easier to work with if the `output` containers are initialized inside of the function AND have proper types, eg, you know that the certain fields will always be string, etc, so we do not need to use the generic containers with "Any" type, eg, `Dict()` is effectively `Dict{Any,Any}()` -> We probably know more about its types?
> 
> If you want to have `process_nodes` incrementally fill these output containers, the convention is to put "!" in the function name to indicate that it's a mutating function -> `process_nodes!` - see https://docs.julialang.org/en/v1/manual/style-guide/#bang-convention

So in `Dict`, the key is always `String` so I'll specify `String` but value could either be a `String` or a `Dict`, so should I use `Any`?

**Comment by splendidbug** [comment]
> @splendidbug How are you thinking about the scope of this PR before we merge it?
> 
> I think we can have it open for a while, but I'd like to see the following prior to merge:
> 
>     * functionality:
>       
>       * download web page (can be in memory, or saved to a local file for future reference/re-parsing, it should reflect the webpage structure. HTML file only)
>       * parsing HTML page
>       * extracting links from HTML page
>       * filtering pages to only the same domain(?, should be an argument
>       * top-level functionality to run the scrape->links->parse & repeat on other links in scope
> 
>     * tests for the main behaviours
> 
> 
> It doesn't have to cover all edge cases, but it would be good to cover all these steps, because it will force us to design a good interface/API from the start (and not change it soon).
> 
> Eg, in separate PRs we should consider checking anti-scraping measures, eg, robots.txt, and make sure we adhere to it. But that's extra and not the core scope of this PR.
> 
> WDYT?

Sounds good and I agree that keeping the PR open to add other functionalities is a good idea
Regarding, filtering pages to only the same domain, isn't that necessary to avoid virtually infinite urls? Is there a use-case where it's beneficial to go out of domain?

Also when we start implementing the scraper, we'll have to take care of memory overflow right?

**Comment by svilupp** [comment]
> Regarding, filtering pages to only the same domain, isn't that necessary to avoid virtually infinite urls? Is there a use-case where it's beneficial to go out of domain?

Yes, we will always need to have some filter. The value in having it user provided when required is that you will be able to capture multidoc sites, ie, sites that cover multiple packages (so you would want to filter against a list of domains). Example is sciml.ai I think


> Also when we start implementing the crawler, we'll have to take care of memory overflow right?
Interesting. Why is that a concern for you?
Assuming scraping one site at a time, we shouldn’t get anywhere near RAM limits on a standard laptop (I’d expect MBs of data at most. 
We can add serialization step when we run the scraper for 10000s websites in a loop, but that won’t happen and user can do it themselves if they have such huge task.

Or did you have a different concern?

**Comment by svilupp** [comment]
Separately, could you please go through the review and click resolve on the feedback you tackled already? Also, there were some suggestions - in the future, you can just accept those and it will make the changes for you :)

**Comment by splendidbug** [comment]
> > Regarding, filtering pages to only the same domain, isn't that necessary to avoid virtually infinite urls? Is there a use-case where it's beneficial to go out of domain?
> 
> Yes, we will always need to have some filter. The value in having it user provided when required is that you will be able to capture multidoc sites, ie, sites that cover multiple packages (so you would want to filter against a list of domains). Example is sciml.ai I think
> 
> > Also when we start implementing the crawler, we'll have to take care of memory overflow right?
> > Interesting. Why is that a concern for you?
> > Assuming scraping one site at a time, we shouldn’t get anywhere near RAM limits on a standard laptop (I’d expect MBs of data at most.
> > We can add serialization step when we run the scraper for 10000s websites in a loop, but that won’t happen and user can do it themselves if they have such huge task.
> 
> Or did you have a different concern?

That was my concern. Thanks!

File: DocsScraper/src/DocsScraper.jl [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            end
+
+            # changing `text_to_insert` to "" to avoid inserting `text_to_insert` again (as it was inserted by the child recursion call)
+            if is_text_inserted
+                text_to_insert = ""
+                prev_text = ""
+            end
+
+            # to separate the code into serparate blocks if the code is large and part of text on the webpage
+            if !isempty(strip(recieved_text)) && is_code == true && length(recieved_text) > 50
```
Comment (svilupp) [inline DocsScraper/src/DocsScraper.jl:105]: Minor comment - I've already marked this in the past. Could you please remove the typo `recieved_*`?

File: DocsScraper/src/DocsScraper.jl [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+# Arguments
+- `node`: The root HTML node 
+- `args[1]`: heading_hierarchy - Dict used to store metadata
+- `args[2]`: parsed_text - Vector of Dicts to store parsed text and metadata
+- `args[3]`: child_new - Bool to specify if the current block (child) is part of previous block or not. 
+                        If it's not, then a new insertion needs to be created in parsed_text
+- `args[4]`: prev_text - String which contains previous text
+"""
+function process_node!(node::Gumbo.HTMLElement, args...)
+    heading_hierarchy = length(args) >= 1 ? args[1] : Dict{String,Any}()
```
Comment (svilupp) [inline DocsScraper/src/DocsScraper.jl:43]: Can we please avoid the unpacking of args like that? It's harder to read and harder to test.

I'd suggest having multiple function definitions and they should be just forwarding to some low level implementation (that's the standard pattern)

File: DocsScraper/src/DocsScraper.jl [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        # Recursively process the child node for text content
+        children = collect(AbstractTrees.children(node))
+        num_children = length(children)
+
+        for (index, child) in enumerate(children)
+
+            # if the current tag belongs in the list, it is assumed that all the text/code should be part of a single paragraph/block, unless,
+            # there occurs a code block with >50 chars, then, previously parsed text is inserted first, then the code block is inserted. 
+            if tag_name in ["p", "li", "dt", "dd", "pre", "b", "strong", "i", "cite", "address", "em", "td"]
```
Comment (svilupp) [inline DocsScraper/src/DocsScraper.jl:92]: Could we simplify the code backtick handling by focusing on the PARENT (`parent(node)`) attribute once we get to the `code` block?

The simple rule I see is to decide the code separator when we get the code block based on its parent, eg,
```
function code_fences(node::HTMLElement)
  
  if parent(tag(node)) == :pre
    # implicitly assumes julia, can be easily improved by checking the code CLASS attribute
    "```julia", "```"
  else
    "`", "`"
  end
end

open_fence,close_fence = code_fences(node)
# mock implementation
chunk = open_fence * content * close_fence
```




I'm referring to: https://github.com/JuliaWeb/Gumbo.jl/blob/afc2b2b83501d483e416d86063d98d567968fea7/src/htmltypes.jl#L15
To make your life easy, export it on module level `using AbstractTrees: parent, children` (to have easily available).

Comment (svilupp) [inline DocsScraper/src/DocsScraper.jl:92]: Can you please use the symbol references for tag iteration? It's cleaner and much better for GC (Symbols do not affect GC, strings do and they repeat a lot so it's wasteful)

File: DocsScraper/src/DocsScraper.jl [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+# Arguments
+- `node`: The root HTML node 
+- `args[1]`: heading_hierarchy - Dict used to store metadata
+- `args[2]`: parsed_text - Vector of Dicts to store parsed text and metadata
+- `args[3]`: child_new - Bool to specify if the current block (child) is part of previous block or not. 
+                        If it's not, then a new insertion needs to be created in parsed_text
+- `args[4]`: prev_text - String which contains previous text
+"""
+function process_node!(node::Gumbo.HTMLElement, args...)
```
Comment (svilupp) [inline DocsScraper/src/DocsScraper.jl:42]: For readability purpose, it would be great to:
- move the implementation to a separate script file in `src/` folder
- break down this definition to make it easier to follow (ie, increase the level of abstraction)
   I'd suggest defining `process_node!` for all key node types in the logic (more modular approach and easier to follow), eg, `process_node!(type::CodeBlock(), node, ...)` vs `process_node!(type::TextBlock, node, ...)`
    Define the core block types as: `abstract type AbstractParsedBlock end; struct CodeBlock<:AbstractParsedBlock end;` etc.

**Review by svilupp — CHANGES_REQUESTED** [review]
Good stuff! Some minor changes requested to add more modular functions and to simplify the logic

**Comment by svilupp** [comment]
Btw as mentioned on Slack, an easy way to accumulate strings is to pass around an `io=IOBuffer()` which is `io::IO` in the `process_node` function, that the child nodes can add into (eg, `print(io,…)` or `write(io,…)`) I prefer print somehow - it avoids having to instantiate all the intermediate strings..
You extract them from io via `str = String(take!(io))` (which resets the positions in io / removes all its content, ie, you can do it only once)

**Comment by svilupp** [comment]
As discussed, I haven't looked at the parser yet.

One small observation - DocsScraper should be its own module and only importing/exporting things. No code definitions in there.
Ie,
```julia
module DocsScraper

using AbstractTrees
using Gumbo
using HTTP
using URIs

export x,y,z
include("parser.jl")

export x,y,z
include("...")

end
```

**Comment by svilupp** [comment]
On a separate note, I took it for a spin and parse docs across several packages -- I haven't verified all in detail, but at least it runs across several docsite types (it required small tweaks).

It integrated nicely into the PromptingTools RAGTools:
```julia
## Load up all Makie docs
dirs = ["makie/Makie.jl-gh-pages/dev",
    "makie/AlgebraOfGraphics.jl-gh-pages/dev",
    "makie/GeoMakie.jl-gh-pages/dev",
    "makie/GraphMakie.jl-gh-pages/dev",
    "makie/MakieThemes.jl-gh-pages/dev",
    "makie/TopoPlots.jl-gh-pages/dev",
    "makie/Tyler.jl-gh-pages/dev"
]
output_chunks = Vector{SubString{String}}()
output_sources = Vector{String}()

for dir in dirs
    dir = dirs[1]
    @info ">> Directory: $dir"
    files = mapreduce(x -> joinpath.(Ref(x[1]), x[3]), vcat, walkdir(dir))
    files = filter(x -> endswith(x, ".html"), files)
    chunks, sources = RT.get_chunks(DocParserChunker(), files)
    append!(output_chunks, chunks)
    append!(output_sources, sources)
end

length(output_chunks), length(output_sources)
```

I can share the full script of methods added, but what's relevant is probably this.
You probably recognize your HTML parser. I had to add support for Documenter, Franklin, VitePress (each catching slightly different HTML object as "content" node). Documented in the comments.

```julia
## HTML parser from txt -> vector of dict (ie, skips the download)
"Parses an HTML string into a vector of Dicts with text and metadata. Returns: `parsed_blocks` and `title` of the document."
function parse_html_to_blocks(txt::String)
    parsed_blocks = Vector{Dict{String,Any}}()
    heading_hierarchy = Dict{Symbol,Any}()

    r_parsed = parsehtml(txt)

    # Getting title of the document 
    title = [el
             for el in AbstractTrees.PreOrderDFS(r_parsed.root)
             if el isa HTMLElement && tag(el) == :title] .|> text |> Base.Fix2(join, " / ")

    # Content markers:
    # Documenter: div:docs-main, article: content (within div:#documenter)
    # Franklin: div:main -> div:franklin-content (within div:#main)
    # Vitepress: div:#VPContent

    ## Look for element ID (for Vitepress only)
    content_ = [el
                for el in AbstractTrees.PreOrderDFS(r_parsed.root)
                if el isa HTMLElement && getattr(el, "id", nothing) in ["VPContent"]]
    if length(content_) == 0
        ## Fallback, looking for a class
        content_ = [el
                    for el in AbstractTrees.PreOrderDFS(r_parsed.root)
                    if el isa HTMLElement && getattr(el, "class", nothing) in ["content", "franklin-content"]]
    end

    if length(content_) > 0
        process_node!(only(content_), heading_hierarchy, parsed_blocks)
    end

    return parsed_blocks, title
end

```

This is not a reference implementation, just a quick hack to test it out. I'm saving here for future reference.

**Comment by splendidbug** [comment]
> As discussed, I haven't looked at the parser yet.
> 
> One small observation - DocsScraper should be its own module and only importing/exporting things. No code definitions in there. Ie,
> 
> ```julia
> module DocsScraper
> 
> using AbstractTrees
> using Gumbo
> using HTTP
> using URIs
> 
> export x,y,z
> include("parser.jl")
> 
> export x,y,z
> include("...")
> 
> end
> ```

Gotcha, will make changes
